packages/evm/contracts/adapters/Proofcast/ProofcastAdapter.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.20;
+
+import { Message } from "../../interfaces/IMessage.sol";
+import { MessageIdCalculator } from "../../utils/MessageIdCalculator.sol";
+import { MessageHashCalculator } from "../../utils/MessageHashCalculator.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ECDSA } from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import { RLPReader } from "solidity-rlp/contracts/RLPReader.sol";
+import { IERC777Recipient } from "@openzeppelin/contracts/interfaces/IERC777Recipient.sol";
+import { IERC1820RegistryUpgradeable } from "@openzeppelin/contracts-upgradeable/interfaces/IERC1820RegistryUpgradeable.sol";
+import { BlockHashAdapter } from "../BlockHashAdapter.sol";
+
+contract ProofcastAdapter is BlockHashAdapter, MessageIdCalculator, MessageHashCalculator, Ownable {
+    string public constant PROVIDER = "proofcast";
+
+    // MessageDispatched(uint256 indexed messageId, Message message)
+    bytes32 public constant MESSAGE_DISPATCHED_EVENT_TOPIC =
+        0x218247aabc759e65b5bb92ccc074f9d62cd187259f2a0984c3c9cf91f67ff7cf;
+    uint256 public constant TEE_ADDRESS_CHANGE_GRACE_PERIOD = 48 hours;
+
+    address public teeAddress;
+    address public teeAddressNew;
+    uint256 public teeAddressChangeGraceThreshold;
+    mapping(uint256 => address) public yahos;
+
+    error NotUsableYaho();
+    error InvalidEventRLP();
+    error InvalidTeeSigner();
+    error InvalidSignature();
+    error InvalidYahoAddress();
+    error InvalidEventContentLength(uint256 length);
+    error UnsupportedProtocolId(bytes1 protocolId);
+    error UnsupportedChainId(uint256 chainId);
+    error UnexpectedEventTopic(bytes32 topic);
+    error InvalidSender();
+    error InvalidMessageId(uint256 actual, uint256 expected);
+    error InvalidDestinationChainId(uint256 chainId);
+    error GracePeriodNotElapsed();
+    error InvalidNewTeeSigner();
```

Comment:
Out of completeness you could include a test for this new error type as well, although it is not necessary, as `applyNewTeeSigner` is very straight forward.